pkg/server/sessions/manager.go

```diff
@@ -0,0 +1,184 @@
+
+var guard *manager
+
+type manager struct {
+	Running    bool
+	sessionMux sync.Mutex
+	sessions   map[string]*Session
+	ticker     *time.Ticker
+	done       chan bool
+	logger     logger.Logger
+	options    *Options
+}
+
+type Manager interface {
+	SessionPresent(sessionID string) bool
+	AddSession(sessionID string, sess *Session)
+	RemoveSession(sessionID string)
+	UpdateSessionActivityTime(sessionID string)
+	UpdateHeartBeatTime(sessionID string)
+	StartSessionsGuard() error
+	StopSessionsGuard() error
+	GetSession(sessionID string) *Session
+	CountSession() int
+}
+
+func NewManager(options *Options) *manager {
+	if options == nil {
+		options = DefaultOptions()
+	}
+	guard = &manager{
+		sessionMux: sync.Mutex{},
+		sessions:   make(map[string]*Session),
+		ticker:     time.NewTicker(options.SessionGuardCheckInterval),
+		done:       make(chan bool),
+		logger:     logger.NewSimpleLogger("immudb session guard", os.Stdout),
+		options:    options,
+	}
+	return guard
+}
+
+func (sm *manager) SessionPresent(sessionID string) bool {
+	sm.sessionMux.Lock()
+	defer sm.sessionMux.Unlock()
+	if _, ok := sm.sessions[sessionID]; ok {
+		return true
+	}
+	return false
+}
+
+func (sm *manager) AddSession(sessionID string, sess *Session) {
+	sm.sessionMux.Lock()
+	defer sm.sessionMux.Unlock()
+	sm.sessions[sessionID] = sess
+	sm.logger.Debugf("created session %s", sessionID)
+}
+
+func (sm *manager) GetSession(sessionID string) *Session {
+	sm.sessionMux.Lock()
+	defer sm.sessionMux.Unlock()
+	return sm.sessions[sessionID]
+}
+
+func (sm *manager) RemoveSession(sessionID string) {
+	sm.sessionMux.Lock()
+	defer sm.sessionMux.Unlock()
+	delete(sm.sessions, sessionID)
+}
+
+func (sm *manager) UpdateSessionActivityTime(sessionID string) {
+	sm.sessionMux.Lock()
+	defer sm.sessionMux.Unlock()
+	if sess, ok := sm.sessions[sessionID]; ok {
+		sess.lastActivityTime = time.Now()
+		sm.logger.Debugf("updated last activity time for %s", sessionID)
+	}
+}
+
+func (sm *manager) UpdateHeartBeatTime(sessionID string) {
+	sm.sessionMux.Lock()
+	defer sm.sessionMux.Unlock()
+	if sess, ok := sm.sessions[sessionID]; ok {
+		sess.lastHeartBeat = time.Now()
+		sm.logger.Debugf("updated last heart beat time for %s", sessionID)
+	}
+}
+
+func (sm *manager) CountSession() int {
+	sm.sessionMux.Lock()
+	defer sm.sessionMux.Unlock()
+	return len(sm.sessions)
+}
+
+func (sm *manager) StartSessionsGuard() error {
+	sm.sessionMux.Lock()
+	if sm.Running == true {
+		return ErrGuardAlreadyRunning
+	}
+	sm.Running = true
+	sm.sessionMux.Unlock()
+	for {
+		select {
+		case <-sm.done:
+			return nil
+		case <-sm.ticker.C:
+			sm.expireSessions()
+		}
+	}
+}
+
+func (sm *manager) StopSessionsGuard() error {
+	sm.sessionMux.Lock()
+	defer sm.sessionMux.Unlock()
+	if sm.Running == false {
+		return ErrGuardNotRunning
+	}
+	sm.Running = false
+	sm.ticker.Stop()
+	sm.done <- true
+	sm.logger.Debugf("shutdown")
+	return nil
+}
+
+func (sm *manager) expireSessions() {
+	if sm.Running {
```

Comment:
Access to all fields should be protected by a mutex here, otherwise we'll have race conditions (even the sm.Running boolean).

for the best performance I'd 1) lock the mutex 2) check if we're still running 3) iterate over all sessions, update state accordingly and gather DEAD sessions 4) unlock mutex 5) cleanup dead sessions e.g. rollback pending transactions - I'd avoid doing this step with locked mutex

Reply:
I would like to modify this in transaction handling PR, since it's something could be adapted during implementation